docs/cpp_docs.rst

```diff
@@ -26,94 +44,556 @@ The beginning of both header and source code files should include
+       */
 
-File Description Comment for Header Files
-=========================================
+.. _cpp-doxygen-basics:
+
+Basic Format of Documentation Blocks
+=====================================
+
+This page covers the most important concepts in writing Doxygen-friendly C++ comments for the LSST Stack, though `Doxygen's manual <http://www.doxygen.org/manual/>`_ is the most comprehensive reference.
+
+.. _cpp-doxygen-javadoc:
+
+Documentation MUST be delimited in Javadoc style
+-------------------------------------------------
+
+Documentation blocks **must** begin with ``/**`` and end in ``*/``. The only exception is for :ref:`inline parameter documentation <cpp-doxygen-parameters-inline>`, which may begin with ``///<``.
 
-After the boilerplate, any C++ header file should have a Doxygen-style comment block (that begins with ``/**``) that includes the following fields:
+For consistency, do not begin documentation blocks with ``/*!``.
 
-1. ``@file`` giving the file's name
+.. _cpp-doxygen-tag:
 
-2. ``@brief`` to provide a synopsis of the file for Doxygen's index. This should be a line, at most.
+Documentation MUST use Javadoc-style tags
+------------------------------------------
 
-3. ``@ingroup`` to specify the name of the LSST Stack package containing this file (see :ref:`below <doc-cpp-package-definition>`).
+Documentation blocks must use tags such as ``@see`` or ``@param`` in place of ``\see`` or ``\param``.
+This is both for internal consistency and to avoid conflicts with other tools that give special treatment to ``\word``.
 
-4. ``@author`` to provide the name of the file's primary author.
+.. _cpp-doxygen-form:
+
+Documentation delimiters SHOULD be on their own lines
+------------------------------------------------------
+
+The block's summary sentence occurs on the line after the opening ``/**``.
+
+The terminating ``*/`` should be on its own line, even for "one-line" documentation. For example, a one-line comment:
+
+.. code-block:: cpp
+
+   /**
+    * Sum numbers in a vector.
+    */
+
+(*Note:* one-line documentation blocks are rarely used for public APIs, see :ref:`cpp-doxygen-sections`.)
+
+An example of multi-paragraph documentation:
+
+.. code-block:: cpp
+
+   /**
+    * Sum numbers in a vector.
+    *
+    * This sum is the arithmetic sum, not some other kind of sum that only
+    * mathematicians have heard of.
+    *
+    * @param values Container whose values are summed.
+    * @return sum of `values`, or 0.0 if `values` is empty.
+    */
+
+.. _cpp-doxygen-refs:
+
+Cross-References MUST be prefixed by `@ref`, and SHOULD be used sparingly
+--------------------------------------------------------------------------
+
+References to other classes, functions, and methods will only be linked if they are prefixed by the ``@ref`` tag. Excessive use of links can make the compiled documentation difficult to read; it is recommended that only the first mention of a particular component be a ``@ref``.
```

Comment:
Okay, I see. I think this runs counter to the direction I want to go with our docs (that all API entities are always consistently hyperlinked), and I acknowledge that it'll be a design consideration to make this clean.

I'm fine if we leave this rule in for now, but I may push for amendment when our C++ API docs are run through Sphinx and I'm developing the visual presentation of C++ docs.